type-check.ts

```diff
@@ -351,14 +380,36 @@ export function tcExpr(env: GlobalTypeEnv, locals: LocalTypeEnv, expr: Expr<null
           return tConstruct;
         }
       } else if (env.functions.has(expr.name)) {
-        const [argTypes, retType] = env.functions.get(expr.name);
+        const [params, retType] = env.functions.get(expr.name);
+        const argTypes = params.map((p) => p.type);
         const tArgs = expr.arguments.map((arg) => tcExpr(env, locals, arg));
 
         if (
           argTypes.length === expr.arguments.length &&
           tArgs.every((tArg, i) => tArg.a === argTypes[i])
         ) {
           return { ...expr, a: retType, arguments: expr.arguments };
+        }
+        // case where the function may have default values
+        else if (
+          argTypes.length > expr.arguments.length &&
+          tArgs.every((tArg, i) => tArg.a === argTypes[i])
+        ) {
+          // check if arguments less than number of parameters
+          // first populate all the arguments first.
+          // Then, populate the rest of the values with defaults from params
+          var augArgs = expr.arguments;
+          var argNums = expr.arguments.length;
+          while (argNums < argTypes.length) {
+            if (params[argNums].value === undefined) {
+              throw new Error("Missing argument from call");
+            } else {
+              // add default values into arguments as an Expr
+              augArgs = augArgs.concat({ tag: "literal", value: params[argNums].value });
```

Comment:
Interesting! So this is really restricted to literals. I think a totally different strategy will be needed for non-literals, because we can't copy those expressions around.

Reply:
@DavidLiu0304 @danielytan 
CC @DDEle because Joe's suggestion completely reverses the restriction what we decided earlier.

I would like some clarification on your suggestion above.
So, our goal is to enable not just literals, but expressions for default arguments. Let's use an example for this:
```
class Bar(object):
    baz : int = 0

def f(foo : int, bar : Bar()) -> int:
    return foo + bar.baz

print(f(7))
```

If I'm understanding this correctly, in the above example, `foo` would be the aforementioned uninitialized parameter. `foo` will be tagged with the hypothetical "uninitialized parameter" tag. This would just be adding a new tag to the AST, right?
```
We could make up a new tag for a value that means “uninitialized parameter.” Then, in the body of the function, we could add statements at the beginning that check if a parameter is uninitialized, and if it is, assign the parameter to the correct initial value.
```
I'm not sure what this means. Above, `foo` is uninitialized, but how would we assign the parameter to the correct initial value? In the compiler, don't we already do that just by putting arguments on the stack and make the call? What about overriding a default value by providing an argument to a parameter?

Reply:
From what I envision, in compiler.ts, when we encounter an "uninitialized parameter" in a function call, we would generate the code for that parameter much like the code generation for "construct". Then, we could use that value in the argument for the function call.

Reply:
@jpolitz 

I had some discussion with @DavidLiu0304 and @danielytan again, and we just realized that we don't have to be just restricted to literals for default arguments. 
One point I brought up during our meeting is that, in the parser, if we change traversing for literal to traversing for expression, then can't we just use the existing methodology to also do something like:
```
def func(a : foo = Foo()):
  return a.x
func()
```
In other words, in the existing type check, all we have to do is just insert a `construct` expression.
I believed that we previously were not able to do this because we had this restriction in the parser.

Reply:
Expressions in general are interesting here.

If the constructor `Foo()` prints, how many times should we see a print happen if we call `func` 10 times?